### PR TITLE
[75] Prepare user model for IDR

### LINKED ISSUE #75: Clean up user model stuff for IDR
We no longer need a preferred and middle name as that will be handled during import / update from IDR. Also remove the ability to edit / update all attributes except for gender, which we will allow students to override locally.

### PR DESCRIPTION
**Pull request by esoterik**
Resolves #75
  - Removes middle names and preferred names from users
  - Adds gender enum to users

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -12,15 +12,17 @@ class User < ApplicationRecord
...
   enum role: %w(student admin rep)
   enum intent: %w(undeclared on_campus off_campus)
+  enum gender: %w(non-binary female male)
 
-  # Returns the user's preferred first name
+  # Returns the user's preferred name
   #
-  # @return [String] Preferred name if set, otherwise first name.
+  # @return [String] Preferred name
   def name
```
Comment (orenyk) [inline app/models/user.rb]: Woot! I like how we don't have to change things _everywhere_ for this 😄

File: app/models/user.rb [hunk 2]
```diff
@@ -12,15 +12,17 @@ class User < ApplicationRecord
   validates :first_name, presence: true
   validates :last_name, presence: true
   validates :intent, presence: true
+  validates :gender, presence: true
 
   enum role: %w(student admin rep)
   enum intent: %w(undeclared on_campus off_campus)
+  enum gender: %w(non-binary female male)
```
Comment (orenyk) [inline app/models/user.rb]: According to the data dictionary, the values we're getting from IDR are:

> Valid values are NULL (unknown), "N" (Not Declared), "F" (Female) or "M" (Male)

How should we handle "N" and NULL cases? "N" doesn't necessarily feel like non-binary to me, but you probably have a better idea about this than I do.

File: spec/factories/users.rb [hunk 3]
```diff
@@ -5,6 +5,7 @@
     password { 'passw0rd' }
     first_name { 'First' }
     last_name { 'Last' }
+    gender { 0 }
```
Comment (orenyk) [inline spec/factories/users.rb]: Why are you using the enum value and not the string?

File: app/views/users/edit.html.erb [hunk 4]
```diff
@@ -1,10 +1,6 @@
 <h1> Editing <%= @user.name %> </h1>
 <%= simple_form_for @user do |f| %>
-  <%= f.input :email %>
-  <%= f.input :first_name %>
-  <%= f.input :preferred_name %>
-  <%= f.input :middle_name %>
-  <%= f.input :last_name %>
-  <%= f.select :role, collection: User.roles.keys %>
+  <%= f.select :gender, collection: User.genders.keys %>
```
Comment (orenyk) [inline app/views/users/edit.html.erb]: Not that we need to change this now, but we're going to need to add some labels to these forms eventually 😛.

File: spec/features/users/edit_spec.rb [hunk 5]
```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+require 'rails_helper'
+
+RSpec.feature 'User Editing' do
+  let(:user) { FactoryGirl.create(:student) }
```
Comment (orenyk) [inline spec/features/users/edit_spec.rb]: Should we specify the gender to make it clear that it's not already `male`?

**Comment by orenyk** [comment]
Weird, one of my comments is claiming it's on an outdated diff but the change is still there (it's on the factory).

File: app/models/user.rb (hunk 2 above)
Comment (esoterik) [inline app/models/user.rb]: I think defaulting to non-binary is fine

**Comment by esoterik** [comment]
updated the feature test!

File: spec/factories/users.rb [hunk 6]
```diff
@@ -5,6 +5,7 @@
     password { 'passw0rd' }
     sequence(:first_name) { |n| "First_#{n}" }
     last_name { 'Last' }
+    gender { 0 }
```
Comment (orenyk) [inline spec/factories/users.rb]: This is the line that it claimed was outdated; is there a reason why we're using the integer value and not the string here?

Comment (esoterik) [inline spec/factories/users.rb]: nope, switched it to the string

**Comment by orenyk** [comment]
Actually, it looks like that broke some things on Travis...

File: spec/factories/users.rb [hunk 7]
```diff
@@ -5,6 +5,7 @@
     password { 'passw0rd' }
     sequence(:first_name) { |n| "First_#{n}" }
     last_name { 'Last' }
+    gender { 'non_binary' }
```
Comment (orenyk) [inline spec/factories/users.rb]: Oh this is supposed to be `'non-binary'` (dash, not underscore). Feel free to change the Enum definition instead.

**Comment by esoterik** [comment]
sorry about that, should be good to go now

**Pull request merged**